examples/cfq/input_pipeline.py

```diff
@@ -0,0 +1,210 @@
+            datasets=(train_raw,),
+            preprocessing_fun=preprocessing.preprocess_example)
+
+        self.unk_idx = self.vocab[constants.UNK]
+        self.bos_idx = self.vocab[constants.BOS]
+        self.eos_idx = self.vocab[constants.EOS]
+        self.tf_vocab = utils.build_tf_hashtable(self.vocab, self.unk_idx)
+        self.vocab_size = len(self.vocab)
+        self.i2w = list(self.vocab.keys())
+
+        # Turn data examples into pre-processed examples by turning each sentence
+        # into a sequence of token IDs. Also pre-prepend a beginning-of-sequence
+        # token <s> and append an end-of-sequence token </s>.
+
+        self.train_dataset = train_raw.map(
+            self.prepare_example,
+            num_parallel_calls=constants.AUTOTUNE).cache()
+        self.dev_dataset = dev_raw.map(
+            self.prepare_example,
+            num_parallel_calls=constants.AUTOTUNE).cache()
+        self.test_dataset = test_raw.map(
+            self.prepare_example,
+            num_parallel_calls=constants.AUTOTUNE).cache()
+
+
+    def get_padded_shapes(self, output_len):
+        """The padded shapes used by batching functions."""
+        query_pad = None
+        if self.fixed_output_len:
+            query_pad = output_len
+        return {
+            constants.QUESTION_KEY: [None],
+            constants.QUERY_KEY: [query_pad],
+            constants.QUESTION_LEN_KEY: [],
+            constants.QUERY_LEN_KEY: []
+        }
+
+    def add_bos_eos(self, sequence: tf.Tensor) -> tf.Tensor:
+        """Prepends BOS ID and appends EOS ID to a sequence of token IDs."""
+        return tf.concat([[self.bos_idx], sequence, [self.eos_idx]], 0)
+
+    def prepare_sequence(self, sequence: Text):
+        """Prepares a sequence(question or query) by tokenizing it, transforming
+        it to a list of vocabulary indices, and adding the BOS and EOS tokens"""
+        tokenized_seq = self.tf_vocab.lookup(self.tokenizer.tokenize(sequence))
+        wrapped_seq = self.add_bos_eos(tokenized_seq)
+        return wrapped_seq
+
+    def prepare_example(self, example: ExampleType) -> ExampleType:
+        """Prepares an example by converting to IDs and wrapping in <s> and </s>."""
+        example = preprocessing.preprocess_example(example)
+        example[constants.QUESTION_KEY] = self.prepare_sequence(
+            example[constants.QUESTION_KEY])
+        example[constants.QUERY_KEY] = self.prepare_sequence(
+            example[constants.QUERY_KEY])
+        example[constants.QUESTION_LEN_KEY] = tf.size(
+            example[constants.QUESTION_KEY])
+        example[constants.QUERY_LEN_KEY] = tf.size(example[constants.QUERY_KEY])
+        return example
+
+    def ouput_length_fn(self, example: ExampleType) -> tf.Tensor:
+        """Function that takes a dataset entry and returns the query length"""
+        return example[constants.QUERY_LEN_KEY]
+
+    def example_length_fn(self, example: ExampleType) -> tf.Tensor:
```

Comment:
Please rename to `get_example_length`, or (perhaps even clearer) `get_max_example_length`.